src/main/scala/chisel3/experimental/inlinetest/InlineTest.scala

```diff
@@ -114,13 +139,28 @@ trait HasTests[M <: RawModule] { module: M =>
     */
   protected final def test[R](
     testName: String
-  )(testBody: Instance[M] => R)(implicit th: TestHarnessGenerator[M, R]): Unit =
-    elaborateParentModule { moduleDefinition =>
-      val resetType = module match {
-        case module: Module => Some(module.resetType)
-        case _ => None
-      }
-      val test = new TestParameters[M, R](desiredName, testName, moduleDefinition, testBody, resetType)
-      th.generate(test)
+  )(testBody: Instance[M] => R)(implicit testHarnessGenerator: TestHarnessGenerator[M, R]): Unit = {
+    val resetType = module match {
+      case module: Module => Some(module.resetType)
+      case _ => None
     }
+    require(!testGenerators.contains(testName), s"test '${testName}' already declared")
+    val testGenerator =
+      new TestGenerator(
+        index = testGenerators.size,
+        testName,
+        () => module.name,
+        testBody,
+        resetType,
+        testHarnessGenerator
+      )
+    testGenerators.addOne(testName -> testGenerator)
+  }
+
+  afterModuleBuilt {
+    lazy val moduleDefinition = module.toDefinition.asInstanceOf[Definition[M]]
+    testGenerators.values.toSeq.sortBy(_.index).foreach { t =>
+      Definition(t.generate(moduleDefinition))
+    }
```

Comment:
`lazy vals` for function-local values are a bit rare, I can see you're doing this to reuse the object but also to only build it if it's actually used. So if my reading is correct, please say that in a comment.

Alternatively, this trait all provides `moduleDefinition` as a value, can you just use that one?